openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.stop();
+#endif
+    }
+
+    void operator()(const LeafRange& r) const
+    {
+        typename BufferPoolT::reference buffer = mBufferPool->local();
+        static const ValueT Unknown = std::numeric_limits<ValueT>::max();
+        AccT acc(mGrid1->tree());
+        size_t boundaryCount = 0;
+        for (auto leafIter = r.begin(); leafIter; ++leafIter) {// mask leafs
+            // The following hack is safe due to the topoloyUnion in
+            // init and the fact that ValueT is known to be a floating point!
+            ValueT *data = acc.probeLeaf(leafIter->origin())->buffer().data();
+            for (auto voxelIter = leafIter->cbeginValueOn(); voxelIter; ++voxelIter) {// mask voxels
+                if (math::Abs( data[voxelIter.pos()] ) < Unknown ) {
+                    ++boundaryCount;
+                } else {
+                    buffer.push_back(voxelIter.getCoord());
+                }
+            }
+        }
+        mParent->mBoundaryCount += boundaryCount;//reduces pressure on the atomic
+    }
+
+    // Private member data of MaskKernel
+    FastSweeping *mParent;
+    BufferPoolT  *mBufferPool;
+    GridT        *mGrid1;//raw pointer, i.e. lock free
+};// MaskKernel
+
+/// @brief Private class of FastSweeping to perform concurrent fast sweeping in two directions
+template <typename GridT>
+struct FastSweeping<GridT>::SweepingKernel
+{
+    SweepingKernel(FastSweeping &parent) : mParent(&parent),
+      mAcc1(parent.mGrid1->tree()), mTree2(parent.mGrid2?&parent.mGrid2->tree():nullptr),
+      mVoxelSize(static_cast<ValueT>(parent.mGrid1->voxelSize()[0])) {}
+    SweepingKernel(const SweepingKernel&) = default;
+    SweepingKernel& operator=(const SweepingKernel&) = delete;
+
+    /// Main method that performs concurrent bi-directional sweeps
+    template<typename HashOp>
+    void sweep(HashOp hash)
+    {
+#ifdef BENCHMARK_FAST_SWEEPING
+        util::CpuTimer timer("\nConcurrent copy of coordinates");
+#endif
+        Coord *coords = mParent->mCoords.get();
+        mParent->mPagedArray.copy(coords);
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Alternative init");
+#endif
+        assert( mParent->mPagedArray.size() < static_cast<size_t>(std::numeric_limits<uint32_t>::max()) );
+        const uint32_t count = mParent->mPagedArray.size();
+        auto tmp = std::make_unique<uint32_t[]>(count);
+
+        tbb::parallel_for(tbb::blocked_range<uint32_t>(0, count, 64),
+                          [&](const tbb::blocked_range<uint32_t>& r){auto *p=&tmp[r.begin()]; for (uint32_t i = r.begin(); i < r.end(); ++i) *p++=i;});
+        //if (tmp[134] != 134) std::cerr << "ERROR" << std::endl;
+        auto hashComp2 = [&](uint32_t a, uint32_t b){return hash(coords[a]) < hash(coords[b]);};
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Alternative sort");
+#endif
+        tbb::parallel_sort(&tmp[0], &tmp[0] + mParent->voxelCount(), hashComp2);
```

Comment:
Do you need to remove/comment this alternative sorting implementation out for a final merge?

Reply:
good catch - code from an old benchmark test that I forgot to remove.